gobblin-data-management/src/main/java/org/apache/gobblin/data/management/source/LoopingDatasetFinderSource.java

```diff
@@ -188,39 +232,111 @@ protected WorkUnit computeNext() {
         } else {
           WorkUnit workUnit = workUnitForDataset(dataset);
           addDatasetInfoToWorkUnit(workUnit, dataset, this.generatedWorkUnits++);
+          addWatermarkIntervalToWorkUnit(workUnit, dataset);
+          this.previousDataset = dataset;
           return workUnit;
         }
       }
-
-      WorkUnit workUnit = NoopTask.noopWorkunit();
-      workUnit.setProp(WORK_UNIT_ORDINAL, this.generatedWorkUnits);
-
+      WorkUnit workUnit = generateNoopWorkUnit(this.generatedWorkUnits);
       this.generatedWorkUnits = Integer.MAX_VALUE;
-
       workUnit.setProp(END_OF_DATASETS_KEY, true);
       return workUnit;
     }
 
     private void addDatasetInfoToWorkUnit(WorkUnit workUnit, Dataset dataset, int workUnitOrdinal) {
+      if (!Strings.isNullOrEmpty(globalWatermarkDatasetUrn)) {
+        workUnit.setProp(ConfigurationKeys.DATASET_URN_KEY, dataset.getUrn());
+      }
       workUnit.setProp(DATASET_URN, dataset.getUrn());
       workUnit.setProp(WORK_UNIT_ORDINAL, workUnitOrdinal);
     }
 
     private void addPartitionInfoToWorkUnit(WorkUnit workUnit, PartitionableDataset.DatasetPartition partition) {
+      if (!Strings.isNullOrEmpty(globalWatermarkDatasetUrn)) {
+        workUnit.setProp(ConfigurationKeys.DATASET_URN_KEY,
+            Joiner.on(DATASET_PARTITION_DELIMITER).join(partition.getDataset().getUrn(), partition.getUrn()));
+      }
       workUnit.setProp(PARTITION_URN, partition.getUrn());
     }
+
+    private void addWatermarkIntervalToWorkUnit(WorkUnit workUnit, Dataset dataset) {
+      if (!Strings.isNullOrEmpty(globalWatermarkDatasetUrn)) {
+        addWatermarkIntervalToWorkUnit(workUnit, dataset, null);
+      }
+    }
+
+    private void addWatermarkIntervalToWorkUnit(WorkUnit workUnit, Dataset dataset,
```

Comment:
This only works for long watermarks. Would it be better to pass the previously run `WorkUnit` to `workUnitForDataset`/`workUnitForDatasetPartition` for the subclass to decide what to do with it?